### PR TITLE
[WIP][DO NOT REVIEW] make profiling passes a bit more flexible

### PR DESCRIPTION
**Pull request by Krovatkin**
Fixes #{issue number}

### CODE REVIEW

File: torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp [hunk 1]
```diff
@@ -1,4 +1,5 @@
 #include <torch/csrc/jit/runtime/profiling_graph_executor_impl.h>
+#include <torch/csrc/jit/codegen/cuda/parser.h>
```
Comment (jjsjann123) [inline torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp:2]: parser can't be exposed here.
I'll work it out in our API to add a wrapper for it.

File: torch/csrc/jit/codegen/cuda/parser.cpp [hunk 2]
```diff
@@ -651,20 +651,22 @@ std::unordered_map<
...
   pn->insertBefore(node);
-  pn->addInput(in_val);
-  auto pno = pn->addOutput();
-  pno->setType(in_val->type());
-  in_val->replaceAllUsesAfterNodeWith(pn, pno);
-  pn->ty_(attr::profiled_type, in_val->type());
+  // TODO: I think this is better
+  node->replaceInput(offset, pn->output());
+  // TODO: why is this here?
+  // pn->ty_(attr::profiled_type, in_val->type());
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/parser.cpp:668]: I thought we need to know the profiled type as well.
I don't how to retrieved it from `attr::ivalue`.

File: torch/csrc/jit/runtime/profiling_record.cpp [hunk 3]
```diff
@@ -220,6 +159,7 @@ void ProfilingRecord::insertShapeProfile(Node* n, size_t offset) {
 }
 
 bool needsProfiledInputs(Node* n) {
+  // Add CUDAFuser checks
```
Comment (Krovatkin) [inline torch/csrc/jit/runtime/profiling_record.cpp]: @jjsjann123 I couldn't find the equivalent of `tensorexpr::isSupported` 😢

File: torch/csrc/jit/codegen/cuda/parser.cpp (hunk 2 above)
Comment (Krovatkin) [reply torch/csrc/jit/codegen/cuda/parser.cpp:668]: I'd just leave it up to `ProfilingRecord::instrumentBlock` to insert `insertShapeProfileOp`. `prim::profile` is weird and different enough from either `specializeAutogradZero` or `parser.cpp`.

Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/parser.cpp:668]: I was referring to type of scalar value. This is not trying to replace profiling of TensorType.
I was basically using it to determine what scalar to cast the ivalue to.
https://github.com/csarofeen/pytorch/blob/sum_to_size_PR/torch/csrc/jit/codegen/cuda/graph_fuser.cpp#L46-L58

File: torch/csrc/jit/runtime/profiling_record.cpp (hunk 3 above)
Comment (jjsjann123) [reply torch/csrc/jit/runtime/profiling_record.cpp]: Here you go:
https://github.com/csarofeen/pytorch/blob/sum_to_size_PR/torch/csrc/jit/codegen/cuda/register_interface.cpp#L27-L28

`isSupported` should be `canFuseNode`, additional for ivalue profiling, we have `insertProfileIValue`

File: torch/csrc/jit/codegen/cuda/parser.cpp (hunk 2 above)
Comment (Krovatkin) [reply torch/csrc/jit/codegen/cuda/parser.cpp:668]: could we consider giving them different names `attr::profiled_axes`, `attr::cond` ? It looks like we are trying to tell apart axes and a isCuda/isTraining boolean? or are we have an attribute that changes its type?

Comment (Krovatkin) [reply torch/csrc/jit/codegen/cuda/parser.cpp:668]: i can make a change if you think this could work

Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/parser.cpp:668]: that should work.
We might need to add another one later (`attr::profiled_reduction_shapes`) for `_grad_sum_to_size`. Let's not worry about it for now.

**Comment by csarofeen** [comment]
@jjsjann123 is this PR stale? If so can you close it.

**Pull request closed without merging**